crates/wast/src/core/table.rs

```diff
@@ -203,39 +203,65 @@ impl<'a> Parse<'a> for ElemPayload<'a> {
 
 impl<'a> ElemPayload<'a> {
     fn parse_tail(parser: Parser<'a>, ty: Option<RefType<'a>>) -> Result<Self> {
-        let (must_use_indices, ty) = match ty {
+
+        // We are in the middle of parsing an element
+        // (elem XXXX indices/expressions)
+        //       ^^^^
+        //       | we are here. We use the provided context `ty` + parser lookahead to learn which
+        //         abbreviation notation we are currently parsing.
+        match ty {
             None => {
+                // case: "XXXX" is "func", we expect to parse a list
+                // of untyped function references indices.
                 parser.parse::<Option<kw::func>>()?;
-                (true, RefType::func())
+                let mut elems = Vec::new();
+                while !parser.is_empty() {
+                    elems.push(parser.parse()?);
+                }
+                Ok(ElemPayload::Indices(elems))
             }
-            Some(ty) => (false, ty),
-        };
-        if let HeapType::Func = ty.heap {
-            if must_use_indices || parser.peek::<Index<'_>>() {
+            Some(RefType { heap: HeapType::Func, .. }) if parser.peek::<Index<'_>>() => {
+                // case: "XXXX" is the empty string. The context tells
+                // us to expect a list of untyped function references.
                 let mut elems = Vec::new();
                 while !parser.is_empty() {
                     elems.push(parser.parse()?);
                 }
-                return Ok(ElemPayload::Indices(elems));
+                Ok(ElemPayload::Indices(elems))
             }
-        }
-        let mut exprs = Vec::new();
-        while !parser.is_empty() {
-            let expr = parser.parens(|parser| {
-                if parser.peek::<kw::item>() {
-                    parser.parse::<kw::item>()?;
-                    parser.parse()
-                } else {
-                    // Without `item` this is "sugar" for a single-instruction
-                    // expression.
-                    let insn = parser.parse()?;
-                    Ok(Expression {
-                        instrs: [insn].into(),
-                    })
+            Some(ty @ RefType { heap: HeapType::Index(_), .. }) if parser.peek::<Index<'_>>() => {
+                // case: "XXXX" is the empty string. The context tells
+                // us to expect a list of typeful function references.
+                let mut instrs = Vec::new();
+                while !parser.is_empty() {
+                    instrs.push(Instruction::RefFunc(parser.parse()?));
                 }
-            })?;
-            exprs.push(expr);
+                Ok(ElemPayload::Exprs {
+                    ty,
+                    exprs: vec![Expression { instrs:instrs.into() }] })
```

Comment:
> This doesn't feel quite right to me because this is implicitly transforming an index-based list into an expression-based list which is represented differently in the binary format. In the sense that the text format is trying to be a reflection of the binary format I think it would be best if indices remained indices and expressions remained expressions as this also assists in round-tripping a module.

I agree, that's a slight hack. I suppose I can make it more principled by extending the definition of `Indices` with a type component like `Exprs`, and then slightly tweak the binary encoding/decoding as you suggested.

> Reading over the [current online spec for function references](https://webassembly.github.io/function-references/core/text/modules.html#tables) I'm not sure that the elaboration depicted there is possible because the [binary format for elements](https://webassembly.github.io/function-references/core/binary/modules.html#element-section) specifies no means by which function indices can be typed as anything other than `funcref`.

My guess is that this is a typo/oversight. CC @rossberg

Reply:
Oh, seems I was too fast with the "fix" last night, which replaced `func` with `reftype`. Before, the expansion was actually reducing to the abbreviation for element lists [defined here](https://webassembly.github.io/function-references/core/text/modules.html#id7), so that was at least correct, even if surprisingly restrictive.

I can either revert that, or inline the `ref.func` wrapping. Probably will do the latter, which matches the reference interpreter.

Reply:
OK. That makes sense to me. My patch here ought to produce the same binary encoding as the reference interpreter.

Reply:
[Fixed](https://github.com/WebAssembly/function-references/commit/3d09c2201cb694a21fd9bbdfee93fe74aeb5b47b).

Reply:
Thanks. @alexcrichton in light of the spec change, do you want me to change anything in this patch?

Reply:
Ok if that's the direction the spec goes in I think that more changes may be necessary here. The element parsing code here has always been a bit messy, but if this is to strictly follow the spec then all inline elements need to be transformed to lists-of-expressions, regardless of whether it's a typed function reference or not. I believe that list-of-indexes should only be supported now for element segments specifically.

While I think this change may work I think it may unfortunately may also make what's already complicated a bit more complicated, and deviate a bit further from the spec in a few edge cases. In that sense I'd ideally prefer to clean it up before landing if you're up for that? If you'd prefer to not take this on I can also look into cleaning up this element parsing as well.

Reply:
It sounds like you have something concrete in mind already. I am perfectly happy to hand this patch over to you. My agenda is only to get the function references implementation spec complete.

Reply:
Well I don't necessarily want to take this over, but at the same time I think it's worth considering restructuring the code a bit to better handle the predicted spec here rather than adding one case for one test.

Could I perhaps push back a bit on the spec change though? Why should text parsers translate list-of-indices to list-of-expressions? I didn't think that this happened for element segments themselves so is there a reason that this form of inline table elements should be singled out? I understand it's not really a big ask for text parsers to do this but at the same time I'm also not sure myself what the motivation for asking parsers to do this in the first place is other than "the syntax already written down in a test should probably parse"

Reply:
To be clear, it's not a spec change, merely a minor fix to adjust for more precise types, which simply didn't matter before. The syntax itself and its expansion already exist in present Wasm 2.0. Moreover, the same expansion does indeed happen for elem segments themselves, see [here](https://webassembly.github.io/spec/core/text/modules.html#id7).

Reply:
I may be coming at this from a different angle which may be worth correcting if you're up for it. My understanding of the text format is that it's intended to reflect the various structures of the binary format. To that end my expectation is that each binary representation of an element segment would be printed in the text format with a different syntactic structure. For example the [binary format for elements](https://webassembly.github.io/spec/core/binary/modules.html#element-section) lists 8 different encodings for element segments and I'd personally expect that each form would be printed in the text format differently. Furthermore when that text format is parsed it would then reproduce the same binary encoding, avoiding translating between different representations.

In that sense I interpreted the equivalence you noted in your link between list-of-indices and list-of-expressions as more of a semantic equivalence rather than a binary equivalence. Given all this my expectation is that because there is no binary encoding for list-of-indices with a non-`funcref` type then that would not be supported in the text format.

Is the expectation, however, that everything is transformed to list-of-expressions? Or are my expectations of the binary/text format misaligned here?

Reply:
The spec only specifies the _meaning_ of binary and text format. It does so by defining how each is parsed into abstract syntax (which involves some desugaring to reduce spec redundancy). It does not specify anything about the inverse, or about tools doing binary-to-text conversion or back. The reference interpreter simply does the simplest thing for the inverse direction.

Other tools are of course welcome to go for perfect roundtripping (and both formats are almost closely enough aligned to do that, except for some minor coding details like LEBs and local declarations). But that's a tooling question, outside the scope of the spec.

Reply:
Ok well sounds like the spec is staying put. @dhil I don't have a ton of time right now to take this over per se, so if you're up for doing a sort of sync with the current state of the parsing here and with the spec that would be appreciated. 

I'll note though that given the state of the spec I don't think the goal of this implementation will be 100% strict spec complicance because it doesn't make sense for our use case. The goal of the text implementation here is to closely resemble the binary format where it can, so for example the text implementation should not silently translate all list-of-index representations to list-of-expressions.

Reply:
I am happy to make the necessary changes. I will, however, ask you to nudge me in the right direction. So what do we need to change?

Reply:
Well, in general, a once over of the code and corresponding to exising tests and the spec would be good to do. For example this specific case I've highlighted here is not correct as it creates a single expression of many `ref.func` instructions. This only happens to work for the 1-element table showing up in the tests at the moment.

Additionally the comments in this function about what's being parsed aren't quite correct in the context of which this PR is being sent because an element segment isn't being parsed but rather an inline portion of a table definition.

I also think it would be better perhaps to refactor this in to "figure out how to parse" and "then parse" steps. Right now the index parsing is duplicated in two places and it'd be good to deduplicate that.

Finally when this function is originally called there's some nontrivial logic to calculate what type to pass in -- either `None` or `Some` of what was already parsed. I think it would be best to either unconditionally pass it in and rely on logic in this function to figure out what to do, or for comments to be up above explaining the cases for when it's passed in and what effect it has on element segment parsing.

---

Since this is a new previously-unused syntax that's expected to be valid I think it would be good to add more tests as well under the `tests/local/*` folder since the one instance of this in the upstream suite is only a minor slice of what's being added.